File: 프로그래머스/lv3/43238. 입국심사/입국심사.py
def solution(n, times):
    answer = 0
    l = min(times)
    r = max(times)*n
    
    while l<=r:
        m = (l+r)//2
        checked = 0
        
        for time in times:
            checked += m//time
            if checked >= n:
                break
        
        if checked>=n:
            answer = m
            r = m-1
        else:
            l = m+1
    
    return answer

File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.*;

class Solution {
    public long solution(int n, int[] times) {
        long answer = 0;
        int arrlen = times.length;

        long min = 0;
        long max = 1_000_000_000 * (long) n;
        
        while (min <= max) {
            long target = (min + max) / 2;
            long people = 0;
            for (int time : times) {
                people += (target / time);
            }
            
            if (people < n) {
                min = target + 1;
            } else {
                max = target - 1;
                answer = target;
            }
        }
        return answer;
    }
}